components/automate-cli/cmd/chef-automate/restart.go

```
@@ -49,6 +99,164 @@ func runRestartServices(cmd *cobra.Command, args []string) error {
+		errChan <- &ResultWithError{}
+		errChan <- &ResultWithError{}
+		err = getChannelValue(errChan)
+		if err != nil {
+			return err
+		}
+		return handleManagedServices(flags)
+	}
+
+	runRestartCmdForBackend(infra, flags, rs, errChan)
+
+	return getChannelValue(errChan)
+}
+
+func runRestartCmdForFrontEnd(infra *AutomateHAInfraDetails, flags *RestartCmdFlags, rs RemoteCmdExecutor, errChan chan *ResultWithError) {
+	if flags.automate {
+		restartOnGivenNode(flags, AUTOMATE, infra, rs, errChan)
+	} else {
+		errChan <- &ResultWithError{}
+	}
+
+	if flags.chefServer {
+		flags.automate = false
+		restartOnGivenNode(flags, CHEF_SERVER, infra, rs, errChan)
+	} else {
+		errChan <- &ResultWithError{}
+	}
+}
+
+func runRestartCmdForBackend(infra *AutomateHAInfraDetails, flags *RestartCmdFlags, rs RemoteCmdExecutor, errChan chan *ResultWithError) {
+	flags.automate = false
+	flags.chefServer = false
+	if flags.postgresql {
+		restartOnGivenNode(flags, POSTGRESQL, infra, rs, errChan)
+	} else {
+		errChan <- &ResultWithError{}
+	}
+	if flags.opensearch {
+		flags.postgresql = false
+		restartOnGivenNode(flags, OPENSEARCH, infra, rs, errChan)
+	} else {
+		errChan <- &ResultWithError{}
+	}
+}
+
+func restartOnGivenNode(flags *RestartCmdFlags, nodeType string, infra *AutomateHAInfraDetails, rs RemoteCmdExecutor, resultErrChan chan *ResultWithError) {
+	go func(flags RestartCmdFlags, resultErrChan chan *ResultWithError) {
+		nodeMap := constructNodeMapForAllNodeTypes(&flags, infra)
+		cmdResult, err := rs.ExecuteWithNodeMap(nodeMap)
+		resultWithError := &ResultWithError{result: cmdResult, err: err}
+		resultErrChan <- resultWithError
+	}(*flags, resultErrChan)
+}
+
+func getChannelValue(errChan chan *ResultWithError) error {
+	for i := 0; i < 4; i++ {
+		errVal := <-errChan
+		if errVal.err != nil {
+			return errVal.err
+		}
+	}
+	return nil
+}
+
+func handleManagedServices(flags *RestartCmdFlags) error {
+	if flags.postgresql && flags.opensearch {
```

Should return an error in case managed services. Verify and update

---

done